lib/utils/couchclient/src/index.js

```diff
@@ -435,63 +436,54 @@ const couchclient = (part, uri, cons) => {
   });
 
   // Pool of memoized db partition handles
-  // Warning: partitions is a mutable variable, but that's the point of
-  // memoization anyway
-  let partitions = [];
-
-  const poolPartitionDB = (u, db, dbKey) => {
-    // do not pool in-memory DBs
-    if (findWhere([db.dbconsOptions], inMemoryOpt)) {
-      debug('DB handle for %s will not be pooled', puri(u));
-      return;
-    }
+  let partitions = {};
 
-    // Warning: mutating variable partitions
-    // Memoize the db handle with both the read mode and the
-    // requested read/write mode
-    partitions[[u, 'read'].join('-')] = db;
-    partitions[dbKey] = db;
-  };
+  // Get pooled db partition handle(s)
+  const pool = (p, rw, cb) => {
 
-  const createDBHandle = (p, rw, cb) => {
-    const u = dbopt.uri(p);
-    debug('Using db %s in %s mode', puri(u), rw);
-
-    // Return memoized db partition handle or get and memoize a new one
-    // from the given db constructor. DB handles are keyed by db uri and
-    // read/write operating mode
-    const dbKey = [u, rw].join('-');
-    if (partitions[dbKey])
-      return cb(null, partitions[dbKey]);
-
-    debug('Constructing db handle for db %s in %s mode', puri(u), rw);
-    return dbopt.cons(u, {
-      // Skip db setup in read mode, as we don't need the db to be
-      // created if it doesn't exist
-      skip_setup: rw === 'read'
-    }, (err, db) => {
-      if(err) {
-        cb(null, errdb('dbcons-err-' + u, err));
-        return;
-      }
-      poolPartitionDB(u, db, dbKey);
+    // Get a pooled db partition handle for a single partition
+    const pooldb = (p, rw, cb) => {
+      const u = dbopt.uri(p);
+      debug('Using db %s in %s mode', puri(u), rw);
+
+      // Return a memoized db partition handle or get and memoize a new one
+      // from the given db constructor. DB handles are keyed by db uri and
+      // read/write operating mode
+      const dbkey = [u, rw].join('-');
+      lock(dbkey, (err, unlock) => {
```

Comment:
The lock module will always pass `undefined` as error. To ensure that this doesn't regress, I will amend the PR to increase the number of documents used by the corresponding unit test from 100 to 1000. Without proper pooling, this should fail.